src/elements/toniq-slider/toniq-slider.element.ts

```diff
@@ -0,0 +1,407 @@
+        .range {
+            display: flex;
+            height: 8px;
+            width: 100%;
+            border-radius: 4px;
+            margin: 0;
+            margin-bottom: 48px;
+            ${applyBackgroundAndForeground(toniqColors.accentSecondary)};
+        }
+
+        .progress {
+            position: absolute;
+            height: 8px;
+            z-index: 1;
+            pointer-events: none;
+            border-radius: 4px;
+            ${applyBackgroundAndForeground(toniqColors.accentPrimary)};
+        }
+
+        .label {
+            z-index: 1;
+            position: absolute;
+            top: 16px;
+            ${toniqFontStyles.boldParagraphFont};
+            ${noUserSelect};
+        }
+
+        .slider {
+            -webkit-appearance: none;
+            height: 8px;
+            width: 100%;
+            background-color: transparent;
+            margin: 0;
+        }
+
+        #lowerSlider,
+        #upperSlider {
+            position: absolute;
+            left: 0;
+            pointer-events: none;
+        }
+
+        .slider-wrapper {
+            position: relative;
+            width: 100%;
+        }
+
+        .slider::-webkit-slider-thumb {
+            -webkit-appearance: none;
+            position: relative;
+            height: 16px;
+            width: 16px;
+            z-index: 10;
+            cursor: pointer;
+            pointer-events: all;
+            border-radius: 10px;
+            transition: ${interactionDuration};
+            ${applyBackgroundAndForeground(toniqColors.accentPrimary)};
+        }
+
+        .slider::-webkit-slider-thumb:hover {
+            transform: scale(1.2);
+        }
+    `,
+    initCallback: ({props, setProps}) => {
+        setProps({value: props.value < props.min ? props.min : props.value});
+    },
+    renderCallback: ({props, host, dispatch, events}) => {
+        function onSliderResize(
+            entry: Readonly<Pick<ResizeObserverEntry, 'target' | 'contentRect'>>,
+        ) {
+            if (props.double) {
+                fillDoubleRangeSlider();
+            } else {
+                fillRangeSlider(parseInt((entry.target as HTMLInputElement).value));
+            }
+        }
+
+        function fillRangeSlider(value: number) {
+            const slider = host.shadowRoot?.querySelector('.slider') as HTMLInputElement;
+            const progress = host.shadowRoot?.querySelector('.progress') as HTMLElement;
+            const label = host.shadowRoot?.querySelector('.label') as HTMLElement;
+            const sliderWidth = slider.clientWidth;
+
+            const sliderOffset = (8 * value) / props.max;
+            const progressOffset = ((value - props.min) * (0 - 8)) / (props.max - props.min) + 8;
+
+            const progressWidth =
+                (sliderWidth * valueMap(value, props.min, props.max, 0, props.max)) / props.max -
+                sliderOffset;
+            progress.style.width = `${progressWidth + progressOffset}px`;
```

Comment:
Do not imperatively modify the style like this. Instead, add a style attribute to the html element and assign it there:

```TypeScript
`<input style="width: ${progressWidth}"`
```

Reply:
> We can't have the width, left, top and other style set as props

Why not?

Reply:
Maybe it'd be worth adding a `privateProps` field to `element-vir`? 🤔

Reply:
I think that would be great

Reply:
I'm also having a problem, `setProps` takes time to effect, and by the time I'm referencing to an element's position its giving me its position before the `setProps` updates. In screenshot code below I'm setting the progress width via `<input style="width: ${progressWidth}"` using a prop and `setProps` and I needed to set the label position as well to where the thumb is. To do that I need the progress right position. By the time I get `progress.getBoundingClientRect().right` its giving me the right position progress before the `setProps` take effect.
![Screen Shot 2022-06-24 at 10 12 22 AM](https://user-images.githubusercontent.com/99699764/175448457-5b006a65-5bbe-45f6-ae43-408c76b05b03.png)

Reply:
I think this problem would be solved if we have some kind of `nextTick()` where the `setProps` is resolve and set the label position right after. But by doing so the UI would be jerky where the label would always be trailing behind the slider thumb

Reply:
Please ignore the above comment. I found a way. The `privateProps` would be a great addition.